lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1434 @@
+      // annotations.
+      if (anno.isClass(resetAnno)) {
+        reset = op->getResult(0);
+        conflictingAnnos.insert({anno, reset.getLoc()});
+        return true;
+      }
+      if (anno.isClass(ignoreAnno)) {
+        anyFailed = true;
+        op->emitError(
+            "'IgnoreFullAsyncResetAnnotation' cannot target wire/node; must "
+            "target module instead");
+        return true;
+      }
+      return false;
+    });
+  });
+  if (anyFailed)
+    return failure();
+
+  // If we have found no annotations, there is nothing to do. We just leave this
+  // module unannotated, which will cause it to inherit a reset domain from its
+  // instantiation sites.
+  if (!ignore && !reset) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "No reset annotation for " << module.getName() << "\n");
+    return success();
+  }
+
+  // If we have found multiple annotations, emit an error and abort.
+  if (conflictingAnnos.size() > 1) {
+    auto diag = module.emitError("multiple reset annotations on module '")
+                << module.getName() << "'";
+    for (auto &annoAndLoc : conflictingAnnos)
+      diag.attachNote(annoAndLoc.second)
+          << "conflicting " << annoAndLoc.first.getClassAttr() << ":";
+    return failure();
+  }
+
+  // Dump some information in debug builds.
+  LLVM_DEBUG({
+    llvm::dbgs() << "Annotated reset for " << module.getName() << ": ";
+    if (ignore)
+      llvm::dbgs() << "no domain\n";
+    else if (auto arg = reset.dyn_cast<BlockArgument>())
+      llvm::dbgs() << "port " << getModulePortName(module, arg.getArgNumber())
+                   << "\n";
+    else
+      llvm::dbgs() << "wire "
+                   << reset.getDefiningOp()->getAttrOfType<StringAttr>("name")
+                   << "\n";
+  });
+
+  // Store the annotated reset for this module.
+  assert(ignore || reset);
+  annotatedResets.insert({module, reset});
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Domain Construction
+//===----------------------------------------------------------------------===//
+
+/// Gather the reset domains present in a circuit. This traverses the instance
+/// hierarchy of the design, making instances either live in a new reset domain
+/// if so annotated, or inherit their parent's domain. This can go wrong in some
+/// cases, mainly when a module is instantiated multiple times within different
+/// reset domains.
+LogicalResult InferResetsPass::buildDomains(CircuitOp circuit) {
+  LLVM_DEBUG(
+      llvm::dbgs() << "\n===----- Build async reset domains -----===\n\n");
+
+  // Gather the domains.
+  auto instGraph = getAnalysis<InstanceGraph>();
+  auto module = dyn_cast<FModuleOp>(instGraph.getTopLevelNode()->getModule());
+  if (!module) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Skipping circuit because main module is no `firrtl.module`");
+    return success();
+  }
```

Comment:
Is this a limitation of `InstanceGraph` where domain building won't work correctly if a circuit looks like this?

Reply:
I think there was a line in the Scala code that did a similar thing. And generally, since we don't know what exactly the root op in the circuit is, it makes sense to not just make the compiler abort in a `cast<...>`. Also technically a circuit with a extmodule as top is still valid, and InferResets shouldn't just fail in that case.